### ISSUE TITLE
eliminate redundant fields in SubArray

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#38530]
I noticed that `SubArray` has fields `offset1` and `stride1` that are often redundant:
```
julia> dump(@view [3,4,5][1:2])
SubArray{Int64,1,Array{Int64,1},Tuple{UnitRange{Int64}},true}
  parent: Array{Int64}((3,)) [3, 4, 5]
  indices: Tuple{UnitRange{Int64}}
    1: UnitRange{Int64}
      start: Int64 1
      stop: Int64 2
  offset1: Int64 0
  stride1: Int64 1     # REDUNDANT

julia> dump(@view [3,4,5][Base.OneTo(2)])
SubArray{Int64,1,Array{Int64,1},Tuple{Base.OneTo{Int64}},true}
  parent: Array{Int64}((3,)) [3, 4, 5]
  indices: Tuple{Base.OneTo{Int64}}
    1: Base.OneTo{Int64}
      stop: Int64 2
  offset1: Int64 0    # REDUNDANT
  stride1: Int64 1     # REDUNDANT
```
Could we think about eliding those fields depending on the type of `indices`?   This would cut down on storage for arrays of subarrays, for example.

In particular, we could instead store tow fields `_offset1::TO1` and `_stride1::TS1`, whose types are `Nothing` in cases where the values can be inferred from the *type* of `indices`.   The `subarray.offset1` and `subarray.stride1` fields could be emulated by `getproperty` accessors, which would inline to compile-time constants in cases where they are not stored.

Because this would add two new type parameters to `SubArray`, however, I'm not sure if it would be considered backward compatible?

**Comment by timholy** [comment]
For some subarrays, getting rid of those fields would made linear indexing `O(d)` rather than `O(1)` for a `d`-dimensional array. 

Example:
```julia
julia> a = rand(5,5,5,5,5);

julia> v = view(a, 3, 2:2:4, 2, 3, 2);

julia> v.offset1
898

julia> v.stride1
10
```

I am skeptical that this would be a positive step.

**Comment by stevengj** [comment]
@timholy, I'm not proposing getting rid of those fields for *all* subarrays.  Only for those subarrays where they are compile-time constants based on the type of `indices` — primarily 1d subarrays with unit-range indexing.   Hence the additional type parameters.

**Comment by timholy** [comment]
Are you just proposing to leave them undefined via `new` shenanigans in some circumstances? That seems potentially viable. Or some other redesign of the type definition? If the latter, I'm not following you, can you sketch the new type(s)?

**Comment by stevengj** [comment]
Background: This came up for me in the context of [this post](https://discourse.julialang.org/t/way-to-make-sharedarray-over-fixed-length-strings/7082/16?u=stevengj) about making an array of inlined (isbits) short strings using `StringView{SubArray{SVector...}}`.    I noticed that these extra fields force you to store 16 extra bytes per element, which could easily double the size of such an array.

Of course, one could define a new `SubArray`-like type to get around this, but that seems like a waste if the original `SubArray` can be redefined in such a way as to eliminate the redundancy if possible, and I thought perhaps it might have benefits for other performance-critical views.

Would the addition of such type parameters be considered backward-compatible?   My concern is that it would change concrete type declarations of `SubArray`s into abstract declarations.

**Comment by KristofferC** [comment]
>  My concern is that it would change concrete type declarations of SubArrays into abstract declarations.

Yeah, that is the eternal question when it comes to adding/changing type parameters. To what extent are the type parameters part of the public interface. And which ones of them are? Very hard question...

**Comment by timholy** [comment]
No clear answer here, but for perspective I added a type parameter to `ReinterpretArray` during the 1.6 cycle. That fixed some massive (~20x) performance problems, so I felt pretty justified in doing it. A few packages broke, though; ones that had set up custom dispatch on `ReinterpretArray` and ones that had `@assert precompile(...)` statements. (I may be the only person who routinely adds that `@assert`, though, because I want to know when they go stale.)
